clientv3/watch.go

```diff
@@ -61,6 +61,11 @@ func (wr *WatchResponse) Err() error {
 	return nil
 }
 
+// IsProgressNotify returns true if the WatchResponse is progress notification.
+func (wr *WatchResponse) IsProgressNotify() bool {
+	return len(wr.Events) == 0 && !wr.Canceled
```

Comment:
The channel is closed following a `Canceled`. It's there so the client can grab error information before the channel shuts down for good.